src/main/java/org/tugraz/sysds/runtime/data/TensorBlock.java

```diff
@@ -92,15 +271,344 @@ public boolean isEmpty() {
+			destIx[i]++;
+			srcIx[i]++;
+			//calculating next index
+			while (destIx[i] == outBlock.getDim(i)) {
+				destIx[i] = 0;
+				srcIx[i] = offsets[i];
+				i--;
+				if (i < 0) {
+					//we are finished
+					return outBlock;
+				}
+				destIx[i]++;
+				srcIx[i]++;
+			}
+		}
+		return outBlock;
+	}
+
+	public TensorBlock copy(TensorBlock src) {
+		_dims = src._dims.clone();
+		_heterogeneous = src._heterogeneous;
+		if (_heterogeneous) {
+			_basicTensor = null;
+			_dataTensor = src._dataTensor == null ? null : new DataTensorBlock(src._dataTensor);
+		}
+		else {
+			_dataTensor = null;
+			_basicTensor = src._basicTensor == null ? null : new BasicTensorBlock(src._basicTensor);
+		}
+		return this;
+	}
+
+	public TensorBlock copy(int[] lower, int[] upper, TensorBlock src) {
+		if (_heterogeneous) {
+			if (src._heterogeneous) {
+				_dataTensor.copy(lower, upper, src._dataTensor);
+			}
+			else {
+				// TODO perf
+				_dataTensor.copy(lower, upper, new DataTensorBlock(src._basicTensor));
+			}
+		}
+		else {
+			if (src._heterogeneous) {
+				throw new DMLRuntimeException("Copying `DataTensor` into `BasicTensor` is not a safe operation.");
+			}
+			else {
+				_basicTensor.copy(lower, upper, src._basicTensor);
+			}
+		}
+		return this;
+	}
+
+	// `getExactSerializedSize()`, `write(DataOutput)` and `readFields(DataInput)` have to match in their serialized
+	// form definition
+	@Override
+	public long getExactSerializedSize() {
+		// header size (_isDataTensor, _dims.length + _dims[*], isAllocated)
+		long size = 1 + 4 * (1 + _dims.length) + 1;
+		if (isAllocated()) {
+			if (_heterogeneous) {
+				size += _dataTensor._schema.length;
+				for (BasicTensorBlock bt : _dataTensor._colsdata) {
+					if (bt != null)
+						size += getExactBlockDataSerializedSize(bt);
+				}
+			}
+			else {
+				size += 1 + getExactBlockDataSerializedSize(_basicTensor);
+			}
+		}
+		return size;
+	}
+
+	public long getExactBlockDataSerializedSize(BasicTensorBlock bt) {
+		// nnz, BlockType
+		long size = 8 + 1;
+		if (!bt.isSparse()) {
+			switch (bt._vt) {
+				case INT32:
+				case FP32:
+					size += 4 * getLength(); break;
+				case INT64:
+				case FP64:
+					size += 8 * getLength(); break;
+				case BOOLEAN:
+					//TODO perf bits instead of bytes
+					size += getLength(); break;
+					//size += Math.ceil((double)getLength() / 64); break;
+				case STRING:
+					int[] ix = new int[bt._dims.length];
+					for (int i = 0; i < bt.getLength(); i++) {
+						String s = (String) bt.get(ix);
+						size += IOUtilFunctions.getUTFSize(s == null ? "" : s);
+						getNextIndexes(bt.getDims(), ix);
+					}
+					break;
+				case UNKNOWN:
+					throw new NotImplementedException();
+			}
+		}
+		else {
+			throw new NotImplementedException();
+		}
+		return size;
+	}
+
+	@Override
+	public void write(DataOutput out) throws IOException {
+		//step 1: write header (isDataTensor, dims length, dims)
+		out.writeBoolean(_heterogeneous);
```

Comment:
Similar to MatrixBlock, I would recommend to introduce an enum of serialized types EMPTY, BASIC, DATA (or four if needed for schema) and write it our in the first byte - otherwise if gets too spread across the code and thus harder to maintain.